src/simularium/FrameRecorder.ts

```diff
@@ -0,0 +1,144 @@
+ * for download/use, via the handleFile() callback.
+ *
+ *
+ * Note that the VideoCodecs API is unavailable in some browsers, including Firefox,
+ * as of 2/6/2024. Viewport will not call these methods on firefox.
+ *
+ */
+export class FrameRecorder {
+    private getCanvas: () => HTMLCanvasElement | null;
+    private handleFile: (videoBlob: Blob) => void;
+    private encoder: VideoEncoder | null;
+    private muxer?: Muxer<ArrayBufferTarget>;
+    public isRecording: boolean;
+    private frameIndex: number;
+    public supportedBrowser: boolean;
+    private frameRate: number;
+
+    constructor(
+        getCanvas: () => HTMLCanvasElement | null,
+        handleFile: (videoBlob: Blob) => void
+    ) {
+        this.getCanvas = getCanvas;
+        this.handleFile = handleFile;
+        this.encoder = null;
+        this.isRecording = false;
+        this.frameIndex = 0;
+        this.supportedBrowser = !/firefox|fxios/i.test(navigator.userAgent);
+        this.frameRate = DEFAULT_FRAME_RATE;
+    }
+
+    private async setup(): Promise<void> {
+        const canvas = this.getCanvas();
+        if (canvas) {
+            try {
+                // VideoEncoder sends chunks of frame data to the muxer.
+                // Previously made one encoder in the constructor but
+                // making a new one during setup() prevents a bug where
+                // frames returned blank (stale canvas reference?)
+                this.encoder = new VideoEncoder({
+                    output: (chunk, meta) => {
+                        if (this.isRecording && this.muxer) {
+                            this.muxer?.addVideoChunk(chunk, meta);
+                        }
+                    },
+                    error: (error) => {
+                        console.error("Encoder error:", error);
+                    },
+                });
+                const config: VideoEncoderConfig = {
+                    codec: "avc1.420028",
+                    width: canvas.width,
+                    height: canvas.height,
+                };
+                const { supported, config: supportedConfig } =
+                    await VideoEncoder.isConfigSupported(config);
+                if (supported && supportedConfig) {
+                    this.encoder.configure(config);
+                } else {
+                    throw new Error("Unsupported video encoder configuration");
+                }
+                // Muxer will handle the conversion from raw video data to mp4
+                this.muxer = new Muxer({
+                    target: new ArrayBufferTarget(),
+                    video: {
+                        codec: "avc",
+                        width: canvas.width,
+                        height: canvas.height,
+                    },
+                });
+                this.frameIndex = 0;
+            } catch (error) {
+                throw new Error("Error setting up video encoder: " + error);
```

Comment:
this one might be related to the supportedBrowser=false case, but could also include other conditions.